File: C#/s/s/S.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace s
{
    public class S
    {
        private char lineSplit;
        private Encoding encoding;
        private Node<Object> scope;
        public S(char lineSplit,Encoding encoding)
        {
            this.encoding = encoding;
            this.lineSplit = lineSplit;
            scope = library.System.library();
            scope = Node<Object>.kvs_extend("lib-path", LibPath.instance(), scope);
            loadLibKVS(LibPath.instance().calculate("index.lisp"));
        }
        private Object loadValue(String path, bool delay, s.Node<Object> delay_args)
        {
            Object value = s.library.Load.run_e(path, scope, lineSplit,encoding);
            if (delay)
            {
                value = (value as Function).exec(delay_args);
            }
            return value;
        }
        /*将值挂载到指定节点上*/
        private S __loadLibKV__(String key,Object value)
        {
            scope = s.Node<Object>.kvs_extend(key, value, scope);
            return this;
        }
        /// <summary>
        /// 延迟kv
        /// </summary>
        /// <param name="path"></param>
        /// <param name="key"></param>
        /// <param name="delay_args"></param>
        /// <returns></returns>
        public S loadLibKV_delay(String path, String key, s.Node<Object> delay_args)
        {
            return __loadLibKV__(key, loadValue(path, true, delay_args));
        }
        /// <summary>
        /// 不延迟kv
        /// </summary>
        /// <param name="path"></param>
        /// <param name="key"></param>
        /// <returns></returns>
        public S loadLibKV(String path, String key)
        {
            return __loadLibKV__(key, loadValue(path, false, null));
        }
        /*将值作为kvs挂载*/
        private S __loadLibKVS__(Object kvs)
        {
            s.Node<Object> tmp = kvs as s.Node<Object>;
            while (tmp != null)
            {
                String key = tmp.First() as String;
                tmp = tmp.Rest();
                Object value = tmp.First();
                tmp = tmp.Rest();
                scope = s.Node<Object>.kvs_extend(key, value, scope);
            }
            return this;
        }
        /// <summary>
        /// 延迟kvs
        /// </summary>
        /// <param name="path"></param>
        /// <param name="delay_args"></param>
        /// <returns></returns>
        public S loadLibKVS_delay(String path, s.Node<Object> delay_args)
        {
            return __loadLibKVS__(loadValue(path, true, delay_args));
        }
        /// <summary>
        /// 不延迟kvs
        /// </summary>
        /// <param name="path"></param>
        /// <returns></returns>
        public S loadLibKVS(String path)
        {
            return __loadLibKVS__(loadValue(path, false, null));
        }

        public S addDef(String key, Object value)
        {
            scope = Node<Object>.kvs_extend(key, value, scope);
            return this;
        }

        public Object run(String path)
        {
            return s.library.Load.run_e(path, scope, lineSplit, encoding);
        }

        public void shell()
        {
            String cache = "";
            bool come = true;
            QueueRun qr = new QueueRun(scope);
            while (come)
            {
                String tmp = "";
                Console.Write("<=");
                tmp=Console.ReadLine();
                if (tmp == "``")
                {
                    bool will = true;
                    while (will)
                    {
                        tmp = Console.ReadLine();
                        if (tmp == "``")
                        {
                            will = false;
                        }
                        else
                        {
                            cache = cache + tmp + lineSplit;
                        }
                    }
                }
                else
                {
                    cache = tmp;
                }
                if (cache == "exit")
                {
                    come = false;
                }
                else
                {
                    try
                    {
                        Node<Token> tokens = Token.run(cache, lineSplit);
                        cache = "";
                        if (tokens != null)
                        {
                            Exp exp = Exp.Parse(tokens);
                            Object r = null;
                            r = qr.exec(exp);
                            Console.Write("=>");
                            Console.WriteLine(s.library.System.toString(r,true));
                            Console.WriteLine();
                        }
                    }
                    catch (Exception e)
                    {
                        Util.logException(e);
                    }
                }
            }
        }
    }
    public class LibPath : Function
    {
        private LibPath()
        {
            try
            {
                lib_path = Environment.GetEnvironmentVariable("S_LISP");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            if (lib_path == null || lib_path == "")
            {
                lib_path = "D:/S-Lisp/";
            }
            else
            {
                lib_path = lib_path.Replace('\\', '/');
            }
            
            if (!(lib_path[lib_path.Length - 1] == '/'))
            {
                lib_path = lib_path + "/";
            }
        }
        private static LibPath ini=new LibPath();
        public static LibPath instance() { return ini; }
        public String calculate(string path)
        {
            if (!(path[0] == '.'))
            {
                path = "./" + path;
            }
            return Util.absolute_from_relative(lib_path, path);
        }
        private string lib_path="";
        public override object exec(Node<object> args)
        {
            String path=args.First() as String;
            return calculate(path);
        }
        public override string ToString()
        {
            return "lib-path";
        }

        public override Function.FunctionType Function_type()
        {
            return FunctionType.Fun_BuildIn;
        }
    }
}

File: Python/s/S.py
# coding=utf-8
from LocationException import LocationException
from Token import Token
from Exp import Exp
from QueueRun import QueueRun,UserFunction
from System import System
from Function import Function,FunctionType,LibFunction
from Util import Util
from Node import Node

import os

class S:
	"""docstring for S"""
	def __init__(self, lineSplit):
		self.lineSplit = lineSplit
		self.scope=System.library()
		lib_path=LibPath()
		self.scope=Node.kvs_extend("lib-path",lib_path,self.scope)
		self.loadLibKVS(lib_path.calculate("index.lisp"))
		
		
	def loadValue(self,path,delay,delay_args):
		value=Load.run_e(path,self.scope,self.lineSplit)
		if delay:
			value=value.exe(delay_args)
		return value
		
	def __LoadLibKV__(self,key,value):
		self.scope=Node.kvs_extend(key,value,self.scope)
		return self;
	
	def loadLibKV(self,path,key):
		return self.__LoadLibKV__(key,self.loadValue(path,false,None))
	
	def loadLibKV_delay(self,path,key,delay_args):
		return self.__LoadLibKV__(key,self.loadValue(path,true,delay_args))
	
	def __loadLibKVS__(self,kvs):
		tmp=kvs
		while tmp!=None:
			key=tmp.First()
			tmp=tmp.Rest()
			value=tmp.First()
			tmp=tmp.Rest()
			self.scope=Node.kvs_extend(key,value,self.scope)
		return self
	'''
	不延迟KVS
	'''
	def loadLibKVS(self,path):
		return self.__loadLibKVS__(self.loadValue(path, False, None))
	
	'''
	延迟kvs
	'''
	def loadLibKVS_delay(self,path,delay_args):
		return self.__loadLibKVS__(self.loadValue(path, True, delay_args))
	
	def addDef(self,key,value):
		self.scope=Node.kvs_extend(key,value,self.scope)
		return self
	
	def run(self,path):
		return Load.run_e(path, self.scope, self.lineSplit)
	
	def shell(self):
		tmp=""
		cache=""
		come=True
		qr=QueueRun(self.scope)
		while come:
			tmp=raw_input("<=")
			if tmp=="``":
				will=True
				while will:
					tmp=raw_input()
					if tmp=="``":
						will=False
					else:
						cache=cache+tmp+self.lineSplit
			else:
				cache=tmp
				
			if cache=="exit":
				come=False
			else:
				self.__shell__(qr, cache)
				'''
				try:
					self.__shell__(qr, cache);
				except LocationException,lex:
					print(str(lex))
				except BaseException,ex:
					print(ex)
				'''
				cache=""
	def __shell__(self,qr,cache):
		tokens=Token.run(cache,self.lineSplit)
		if tokens!=None:
			exp=Exp.Parse(tokens)
			r=qr.exe(exp)
			print("=>%s" % System.toString(r,True))
		
class LibPath(LibFunction):
	def __init__(self):
		try:
			self.lib_path=os.environ.get("S_LISP")
		except BaseException,be:
			print(be)
		finally:
			if (self.lib_path==None or self.lib_path==""):
				self.lib_path="D:/S_Lisp"
			else:
				self.lib_path=self.lib_path.replace('\\','/')
			if (not self.lib_path[len(self.lib_path)-1]=='/'):
				self.lib_path=self.lib_path+"/"
			#print(self.lib_path)
			
	def calculate(self,path):
		if (not path[0]=='.'):
			path="./"+path
		return Util.absolute_from_relative(self.lib_path,path)
	def exe(self,args):
		path=args.First()
		return self.calculate(path)
	def __str__(self):
		return "lib-path"
	
class Load(LibFunction):
	def __init__(self,baseScope,base_path,lineSplit):
		self.baseScope=baseScope
		self.base_path=base_path
		self.lineSplit=lineSplit
	def __str__(self):
		return "load"
	def exe(self,args):
		r_path=args.First()
		path=Util.absolute_from_relative(self.base_path,r_path)
		return Load.run_e(path,self.baseScope,self.lineSplit)
	
	onLoad=False
	core=None
	@staticmethod
	def run_e(path,scope,lineSplit):
		if Load.onLoad:
			raise BaseException("禁止在加载期间加载")
		else:
			x=Node.kvs_find1st(Load.core,path)
			if x!=None:
				return x.First()
			else:
				Load.onLoad=True
				sb=Util.readTxt(path,lineSplit)
				scope=Node.kvs_extend("load",Load(scope,path,lineSplit),scope)
				scope=Node.kvs_extend("pathOf",PathOf(path),scope)
				tokens=Token.run(sb,lineSplit)
				exp=Exp.Parse(tokens)
				f=UserFunction(exp,scope)
				b=f.exe(None)
				Load.core=Node.kvs_extend(path,Node.extend(b,None),Load.core)
				Load.onLoad=False
				return b
			
class PathOf(LibFunction):
	def __init__(self,basePath):
		self.basePath=basePath
	def __str__(self):
		return "pathOf"
	def exe(self,args):
		if args==None:
			return self.basePath
		else:
			r_path=args.First()
			return Util.absolute_from_relative(self.basePath,r_path)
	
class Parse(LibFunction):
	def __init__(self,defaultScope,lineSplit):
		self.defaultScope=defaultScope
		self.lineSplit=lineSplit
	def __str__(self):
		return "parse"
	def exe(self,args):
		the_str=args.First()
		args=args.Rest()
		scope=self.defaultScope if args==None else args.First()
		tokens=Token.run(str,self.lineSplit)
		exp=Exp.Parse(tokens)
		f=UserFunction(exp,scope)
		return f.exe(None)
	
class CacheValue(Function):
	def __str__(self):
		return "[]"
	def Function_type(self):
		return FunctionType.cache
	def exe(self,args):
		if args==None:
			return self.value
		else:
			self.value=args.First()
			return None
class Cache(LibFunction):
	def __str__(self):
		return "cache"
	def exe(self,args):
		cv=CacheValue()
		cv.exe(args)
		return cv
		
'''
殷富字符串
'''
class Py_Eval(LibFunction):
	def __str__(self):
		return "py-eval"
	def exe(self,args):
		return eval(args.First())
	
'''
调用实例方法
o
key
ps
'''
class Py_Call(LibFunction):
	def __str__(self):
		return "py-call"
	def exe(self,args):
		o=args.First()
		if isinstance(o, basestring):
			o=eval(o)
		args=args.Rest()
		key=args.First()#方法名
		args=args.Rest()
		
		build=Py_Call.run(args)
		ps=build[0]
		return eval("o."+key+build[1])
	@staticmethod
	def run(args):
		ps=[]
		avs=[]
		if args!=None:
			tmp=args.First()
			i=0
			while tmp!=None:
				ps.append(tmp.First())
				avs.append("ps["+str(i)+"]")
				tmp=tmp.Rest()
				i=i+1
			args=args.Rest()
			if args!=None:
				tmp=args.First()
				while tmp!=None:
					key=tmp.First()
					tmp=tmp.Rest()
					value=tmp.First()
					tmp=tmp.Rest()
					ps.append(value)
					avs.append(key+"=ps["+str(i)+"]")
					i=i+1
		return (ps,"("+",".join(avs)+")")
	
'''
调用函数
fun
ps
'''
class Py_Apply(LibFunction):
	def __str__(self):
		return "py-apply"
	def exe(self,args):
		o=args.First()
		if isinstance(o, baststring):
			o=eval(o)
		args=args.Rest()
		build=Py_Call.run(args);
		ps=build[0]
		return eval("o"+build[1]);
	
'''
访问实例属性o.key=value o.key
o
key
[value]如果存在，则设置，不存在，则返回
'''
class Py_Attr(LibFunction):
	def __str__(self):
		return "py-attr"
	def exe(self,args):
		o=args.First()
		if isinstance(o, basestring):
			o=eval(o)
		args=args.Rest()
		key=args.First()
		args=args.Rest()
		if args!=None:
			value=args.First()
			setattr(o,key,value)
			return None
		else:
			return getattr(o,key)
	
'''
设置字典方法o[key]=value o[key]
'''
class Py_Dict(LibFunction):
	def __str__(self):
		return "py-dict"
	def exe(self,args):
		o=args.First()
		if isinstance(o, basestring):
			o=eval(o)
		args=args.Rest()
		key=args.First()
		args=args.Rest()
		if args!=None:
			value=args.First()
			args=args.Rest()
			if args!=None and args.First():#强制更新
				o.update({key:value})
			else:
				o[key]=value
		else:
			return o.get(key,None)
		
class Py_FunctionValue:
	def __init__(self,fun):
		self.fun=fun
	def __call__(self,*args):
		vs=None
		for arg in args:
			vs=Node.extend(arg,vs)
		return self.fun.exe(vs)
'''
将S-Lisp函数转为python函数
'''
class Py_Function(LibFunction):
	def __str__(self):
		return "py-function"
	def exe(self,args):
		fun=args.First()
		return Py_FunctionValue(fun)
#S("\n").shell()		
			
